xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+
+  // Following fields are set only after the ConfigWatcher registered. Once set, they should
+  // never change.
+  @Nullable
+  private ConfigWatcher configWatcher;
+  // The host name portion of "xds:" URI that the gRPC client targets for.
+  @Nullable
+  private String hostName;
+  // The "xds:" URI (including port suffix if present) that the gRPC client targets for.
+  @Nullable
+  private String ldsResourceName;
+
+  XdsClientImpl(
+      // URI of the management server to be connected to.
+      String serverUri,
+      Node node,
+      // List of channel credential configurations for the channel to management server.
+      // Should pick the first supported one.
+      List<ChannelCreds> channelCredsList,
+      SynchronizationContext syncContext,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Stopwatch stopwatch) {
+    this(
+        buildChannel(checkNotNull(serverUri, "serverUri"),
+            checkNotNull(channelCredsList, "channelCredsList")),
+        node,
+        syncContext,
+        timeService,
+        backoffPolicyProvider,
+        stopwatch);
+  }
+
+  @VisibleForTesting
+  XdsClientImpl(
+      ManagedChannel channel,
+      Node node,
+      SynchronizationContext syncContext,
+      ScheduledExecutorService timeService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Stopwatch stopwatch) {
+    this.channel = checkNotNull(channel, "channel");
+    this.node = checkNotNull(node, "node");
+    this.syncContext = checkNotNull(syncContext, "syncContext");
+    this.timeService = checkNotNull(timeService, "timeService");
+    this.backoffPolicyProvider = checkNotNull(backoffPolicyProvider, "backoffPolicyProvider");
+    this.stopwatch = checkNotNull(stopwatch, "stopwatch");
+  }
+
+  @Override
+  void shutdown() {
+    channel.shutdown();
+    shutdownRpcStream();
+  }
+
+  @Override
+  void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
+    checkState(configWatcher == null, "ConfigWatcher is already registered");
+    configWatcher = checkNotNull(watcher, "watcher");
+    this.hostName = checkNotNull(hostName, "hostName");
+    if (port == -1) {
+      ldsResourceName = hostName;
+    } else {
+      ldsResourceName = hostName + ":" + port;
+    }
+    if (rpcRetryTimer != null) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (adsStream == null) {
+      startRpcStream();
+    }
+    adsStream.sendLdsRequest(ldsResourceName);
+  }
+
+  /**
+   * Builds a channel to the given server URI with the first supported channel creds config.
+   */
+  private static ManagedChannel buildChannel(String serverUri,List<ChannelCreds> channelCredsList) {
+    ManagedChannel ch = null;
+    // Use the first supported channel credentials configuration.
+    // Currently, only "google_default" is supported.
+    for (ChannelCreds creds : channelCredsList) {
+      if (creds.getType().equals("google_default")) {
```

Comment:
anyway we can replace the string literal with a const ID defined somewhere? I think this eventually comes from `envoy/api/v2/core/grpc_service.proto` so it will be good to reference where you define this const ID

Reply:
We've been discussing this in a couple of places such as https://github.com/grpc/grpc-java/pull/6362#discussion_r341741797. We will be using "google_default" string literal until future design for channel creds.